Adding ADR-0006

bendk opened the pull request:
This discusses updating our release process so that we get code merged into our consumers more frequently and to prepare us for a monorepo world.

See the ADR for details.

### Pull Request checklist ###
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[ff-android: firefox-android-branch-name]` and/or `[fenix: fenix-branch-name]` to the PR title.

bendk commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+    * On the monthly Firefox merge date, we do roll each branch forward:
+      - Create a new `esr-[version]` branch from `release`, if needed
+      - Merge `beta` into `release` (this should be a fast-forward).
+      - Bump the version number in main to `[firefox-version].0`
```
Should version numbers have 2 components or 3?  I notice android-components just released `108.0.8` should we also have a middle component?  When would that not be 0?

bendk commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+       * Commit a version bump to `[firefox-version].[N+1]`
+       * Make a PR in the corresponding branch for each application.  Create a tool to automate this.
+       * Collaborate with the mobile teams to get the PR merged.
+       * Merge the changes forward into the other branches.  If the change was committed to `release`, then we merge it into `beta`, and `main`.
```
I think this is how things are handled in moz-central, can someone confirm or deny?

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+    * Automate a nightly merge of our code our `main` branch to the `main` branch of the consumer app repos:
+       * Automate an application-services commit that updates our version number to something like `[firefox-version].nightly-[YYYYMMDD]`
+       * For each consumer application:
+         * Create a PR that updates the submodule commit id or vendors the new code in
```
ditto on the submodule problem, it lends itself well to this, but it's possible to do this without it I imagine

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+  * Good, because our code will exist alongside the application code like firefox-android and firefox-ios, which is how it will be in a monorepo world.
+  * Good, because our version numbers will align more closely with the rest of Mozilla.
+
+## Note on breaking changes
+
+Breaking changes to our consumer application code require extra developer work for both the current and new release models.  We should use this time to restructure our code to prevent breaking changes when it makes sense.  This means creating a more stable API boundary between us and our consumers and evolving those API boundaries in a backwards compatible way.  Some examples of this:
+
+  * When we did the recent exception class refactoring, we tried to be intentional about what exceptions we were exposing to consumers.  The goal was to only expose exceptions that were useful and that we expected to be able to support for a while.
+  * The exception reporting rework was done in backward compatible stages: first we added the new handler code and got the applications to start using it.  Then we get the applications to stop using the old code.  Finally, we remove the old code.  Nimbus has also been using techniques like this to evolve their API.
+  * In hindsight, our SQLCipher logins work could have been split into 2 parts.  The first part would have reworked the API of the logins wrapper code to something that we could have supported before the SQLCipher rework, after the rework, and also going forward into the future.  This would introduce breaking changes for our consumers and we could create PRs to discuss and resolve them.  While th ...
```
This section almost wants to be a separate ADR or a document, I think there is a lot of value here about breaking changes and we don't want it overshadowed by the conversation on how we ship

bendk replied:
I was actually planning it to be a separate ADR, but I wasn't sure exactly how to structure that one since it's not so much a decision but more of a recommendation.  I'll give it a try though.

bendk replied:
I ended up just dropping this section because I wasn't sure what to do with it.  It doesn't feel like it belongs in an ADR or a HOWTO to me -- especially since I don't see much value in it.  It seems pretty unlikely that someone is going to read the doc before making a breaking change and decide to restructure their code in some way.

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+      - Merge `main` into `beta` (also a fast-forward).
+      - Start a new changelog header in `main`
+    * Note: these things replace our current release process -- no more cutting releases.
+    * Coordinate this all with release engineering and see if they want to be involved in any of these steps.
```
I'd change the "if" to "how" 😬 I think release engineering should be a major stakeholder in this, although we may do a lot of the work moving towards this model I want to make sure they are fully aware of all of this and the timeline

tarikeshaq replied:
Let's also add mobile to the list of people we should coordinate this with!

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+    * Keep everything the same for a while and wait for other code to move into the monorepo.
+* **(B) Move to a rapid release process where changes in app-services are frequently merged into consumer applications**
+    * Update our branches to match how moz-central: we would have a `main`, `beta` and `release` branch, with nightly being built off of main, as well as `esr-*` branches for older ESR releases.
+    * Change it so our code lives in the repositories for consumer applications including firefox-android and firefox-ios.  We could either use a git submodule or vendor a copy of our code in to accomplish this.  Note: This is already planned for firefox-ios.
```
The submodule problem feels like a separate ADR - can we do all of this proposal currently without the submodule/vendoring?

bendk replied:
> Not that this should impede this work, we could in the interim support both flows or if there's a strong desire for this we can have that conversation again to push the priority on this!

Yes, I think we could support both flows for an interim period.  That said, I'd love to keep that period short.  Maybe we could help out the process or get release engineering to help.


> The submodule problem feels like a separate ADR - can we do all of this proposal currently without the submodule/vendoring?

Yeah, I think so.  I guess that would mean updating our branches, and adding automation to publishing nightly releases and take those releases on the consumer applications.  Maybe that's a better first step, what do others think?  In any case, I'll add this option to the ADR.

bendk replied:
Actually, I'm unsure about this.  The downside of avoiding the submodule/vendoring is that we need to continue to cut releases.  Right now that's a manual job that feels like it takes too much time.  We could automate it, but that's work we'll need to throw away once we start submoduling/vendoring.

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+         * Set up an auto-merge system where if the test suite passes we automatically merge the PR.  As a fallback we could manually do this every day.
+         * Set up an alert that gets sent to `sync-bots` if the test suite fails.  This means we would need to identify why the tests failed and assign a developer to resolve the issue.  The expectation would be that the developer fixes things in time for the next nightly merge.
+    * In order to keep things running smoothly, if a breaking change requires serious review from the consumer teams, the PRs for that change should be approved and ready to merge before that breaking change is merged into application-services.  If we accidentally merge the change into application-services before the PRs are ready, then we should back the change out.
+    * Manually merge our code whenever we make changes to a `beta`, `release`, or `esr-*` branch.
```
I think this section is describing an uplift right?

bendk replied:
Yes this is an uplift.

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+  * Good, because we don't need to spend time cutting releases.
+  * Good, because uplifts can be done much quicker.
+  * Bad, because we need implement new automation.
+  * Bad, because when breaking changes cause the automatic nightly merge to fail for an application.  This means a developer will get the relatively urgent task of resolving things.
```
Are we doing a good/bad based on the baseline? because currently it's even worse.. if anything an automated release means catching breaking changes faster!

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+  * Good, because uplifts can be done much quicker.
+  * Bad, because we need implement new automation.
+  * Bad, because when breaking changes cause the automatic nightly merge to fail for an application.  This means a developer will get the relatively urgent task of resolving things.
+  * Good, because this will unblock future changes from being merged and released on that application.
```
I don't quite understand this point, maybe we can elaborate a bit more?

bendk replied:
The previous 2 points work together, maybe I should merge them into a single bullet somehow.

My main point is that we're going to be losing some flexibility, for example we couldn't do what we did with logins where ios took a long time to take the changes.  However, that flexibility has a huge downside: it makes it really difficult to merge in any new code in the meantime.  So it's both good and bad, but I believe it's good overall.  That's what I was trying to get at with the "We are willing to accept the trade-offs..." decision driver.

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+  * Bad, because we need implement new automation.
+  * Bad, because when breaking changes cause the automatic nightly merge to fail for an application.  This means a developer will get the relatively urgent task of resolving things.
+  * Good, because this will unblock future changes from being merged and released on that application.
+  * Good, because the developer who writes code with a breaking change is responsible for getting the PRs merged to resolve it (rather than the developer who cuts the release).
```
I'm not sure this is covered in this proposal, and I'm not sure we've explicitly defined that in the current world the developer who cuts the release is responsible

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+  * Bad, because when breaking changes cause the automatic nightly merge to fail for an application.  This means a developer will get the relatively urgent task of resolving things.
+  * Good, because this will unblock future changes from being merged and released on that application.
+  * Good, because the developer who writes code with a breaking change is responsible for getting the PRs merged to resolve it (rather than the developer who cuts the release).
+  * Bad, because breaking changes require the corresponding PRs to be approved from all other teams before they can be merged into application-services.
```
heh, this should be the case now too!

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+  * Good, because the developer who writes code with a breaking change is responsible for getting the PRs merged to resolve it (rather than the developer who cuts the release).
+  * Bad, because breaking changes require the corresponding PRs to be approved from all other teams before they can be merged into application-services.
+  * Good, because this is how things will work in a mono-repo world.  Adjusting our workflow now prepares us for the future and will give us insights that may help guide the monorepo migration.
+  * Good, because our code will exist alongside the application code like firefox-android and firefox-ios, which is how it will be in a monorepo world.
```
I'd almost want this to be the separate "submodule" proposal.

tarikeshaq reviewed (commented):
Thanks a ton for taking this on @bendk!!! Nightly builds are cool and I'm excited 🚀 🚀 I'm very supportive of the main concept here, and I had a lot of comments so please let me know if I should clarify anything.


### submodule
I love the idea of submodule in bringing us closer to monorepo! And although it lends itself so well to nightly builds I think it should be separate because:

- Talked with @csadilek about this and he had good points about submodules delaying our move to Mozilla central. Also, if we want to benefit from the submodules by building from source, we'd require devs in the application repos to have Rust tooling installed and we don't want that outside of Mozilla central
- My conversation with @OrlaM concluded with us wanting to wait till Q1/Q2, and I don't want us to wait for nightly builds until then
- It's a step we can come back to after, so it could simplify some this, but I'm not sure


### other teams
I'd love it we can get other teams involved as early as possible while this is still a proposal. I'd love it if @rvandermeulen or someone else from release management can take a look at this, and @csadilek or someone else from mobile (I understand y'all are busy for a holiday reading 😛)

I also think this is worth sending in the RFC list so we get broader approval/visiability

I understand that we might wanna polish this a little before then, but would be good to surface this before landing


### timeline
if this the approach we want go with, it'll be important to break this down and have a timeline so our stakeholders (mobile, release engineering) know what to expect and when. But that seems like a step **after** this document

There's a decent bit of automation that we want to align with mobile's automation so them being involved early on would help us a lot!

bendk commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+## Context and Problem Statement
+
+There are several issues with the current application-services release process:
+  - Our release schedule is relatively infrequent and consumer apps often take a long time to take our releases.  This means that there's a long turnaround time between when code gets merged to our main branch and when it gets released to a nightly build where it can be used and tested.  This conflicts with the rapid release model used by Firefox Desktop, where code changes are frequently release ...
```
A metric would be great here, but I'm not sure how to accurately calculate that.  What if I take a single app-services release that seems representative to me, and measured the time between merge to `application-services/main` and merge to `fenix/main`?

tarikeshaq replied:
Yeah, I like even an anecdote to demonstrate an example

bendk commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+  - Releases require too much developer involvement.  Normal releases take a couple hours, during which a developer needs to monitor the progress and perform several manual steps along the way.  If the unreleased code has breaking changes, then the developer also needs to merge the PR for the consumer application(s) after the release, which is not ideal since that developer is often not the one w ...
+  - The mobile teams are moving towards a weekly release cycle, which may lead to us needing to do more releases for older versions in order to uplift fixes.
+
+At the same time, Mozilla is in the process of moving towards a single monorepo.  In a monorepo, the application-services code will be immediately consumed by all applications as soon as it's merged.  This is drastically different than our current workflow, where code sits in main for long periods of time before being released, then sits in a release for a long period of time before being taken b ...
```
I thought the plan was that iOS, android, and everything else, would eventually end up in moz-central -- although it's going to be a long time and we haven't decided on the specific steps.

bendk replied:
I like the idea of linking to the proposal document, but hesitate because it's a private google doc at this point.  What do others think, link or no link?

tarikeshaq replied:
I vote linking it is OK, this ADR is mostly relevant to Mozilla employees anyways 🤷

bendk commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+    * Update our branches to match how moz-central: we would have a `main`, `beta` and `release` branch, with nightly being built off of main, as well as `esr-*` branches for older ESR releases.
+    * Change it so our code lives in the repositories for consumer applications including firefox-android and firefox-ios.  We could either use a git submodule or vendor a copy of our code in to accomplish this.  Note: This is already planned for firefox-ios.
+    * Automate a nightly merge of our code our `main` branch to the `main` branch of the consumer app repos:
+       * Automate an application-services commit that updates our version number to something like `[firefox-version].nightly-[YYYYMMDD]`
```
It's the application-services version that the applications consume -- i.e. it would replace the `96.1.3` version that we're currently on.   We could also make a github tag, that makes sense to me.  And this would definitely be before the PRs.

bendk commented:
@tarikeshaq that was a really insightful review, thanks so much.  You've convinced me 100% that any submodule/vendoring proposal should be separate and I removed that from the ADR.  In fact, the more I think about it, the less I care about that change.  I think the workflow changes proposed here are much more important.

I restructured the doc based on your comments, then marked a bunch of conversations resolved -- either because I think I addressed them in the new doc, or because I removed a section and they're not relevant anymore.  Feel free to unresolve any of those if I jumped the gun.

bendk commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,83 @@
+          * Create a commit that updates our version number to `[firefox-version].X.[Y+1]`
+          * Add a git tag for the commit based on that version number (`v[firefox-version].X.[Y+1])
+          * Run the `github-release` CI to publish our code to our consumers (publish maven packages, start the `rust-components-swift` CI, etc).
+          * Maybe we should create PRs to merge the changes back forward to other branches.  For example, merging changes from `release` into `beta` and `beta` into `main`.
```
I'd love to get some wisdom from releng for this.  How does this work for moz-central?

bendk commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,85 @@
+        * Automate a nightly release of our `main` branch.
+          * Create a commit that updates our version number to something like `[firefox-version].0.nightly-[YYYYMMDD]`, similar to what `prepare-release.py` does now.
+            * Make sure that the github CI can push this commit without needing PR approval.
+            * The version string is composed of major, minor, and patch components.  This matches firefox-android, but not moz-central, which skips the minor component.  The reason to have a minor component is to support weekly releases for mobile apps.  They may choose to include new features on those apps, which may require us to uplift new features to our release version and that would be a mi ...
```
I'm pretty conflicted about this one, maybe we could just have 2 components like firefox does.  Like I mention below, we would already be breaking from semver.

lougeniaC64 replied:
If we're already breaking from semver we might as well change the version number format imo. Sticking with the 3 component version number might be confusing.

rvandermeulen replied:
I agree that major.minor is all we'd need. By the time we're doing point releases, I can't imagine ever backporting such a significant breaking change that we'd be uncomfortable with more than a minor version bump. The joys of a 4 week release cycle!

rvandermeulen commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,85 @@
+
+Mozilla generally favors a rapid release model to long cycle times.  Firefox desktop adopted rapid releases over a decade ago and there is current work to move both the mobile applications and our services towards rapid releases.  Although there are some costs to rapid releases, as a company we've decided that we're willing to accept them.
+
+Application services should consider moving towards a rapid release model.  The same reasoning that led other teams to adopt rapid releases also applies to us.  Furthermore, the mobile teams are moving towards rapid releases which may depend on us also adopting them.  Finally, we are in the process of [moving towards a single monorepo for android code](https://docs.google.com/document/d/1VMr5rtn- ...
```
> it's somewhat likely that the monorepo will contain the iOS code as will

I haven't heard anything to suggest that we'll eventually consolidate both Android and iOS into one single repo.

rvandermeulen commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,85 @@
+    * Keep everything the same for a while and wait for other code to move into the monorepo.
+* **(B) Move to a rapid release process where changes in app-services are frequently merged into consumer applications**
+    * Phase 1: update our branching model and release process
+        * Update our branches to match how moz-central: we would have a `main`, `beta` and `release` branch, with nightly being built off of main, as well as `esr-*` branches for older ESR releases.
```
Why not use version-specific branches like the mobile repos do? Would avoid the messy merges every cycle.

rvandermeulen replied:
To expand on this, it might be worth approaching this from a similar angle that NSS is currently using: regular updates between NSS tip and m-c during the Nightly cycle (basically in the form of regular A-S nightlies here), then RTM at the end of the Nightly cycle (i.e. creating a vXXX branch off main here), then point releases as needed during beta/release.

rvandermeulen replied:
This is also very close in concept to how pre-monorepo Android Components integrated in Fenix & Focus.

ddurst replied:
@rvandermeulen (sorry, what is "RTM" in this context?)

rvandermeulen replied:
NSS tags the final release (RTM) at the end of the nightly cycle. See https://hg.mozilla.org/projects/nss/tags

bendk replied:
Thanks!  This seems much simpler to me as well.

tarikeshaq reviewed (commented):
I like how this becoming straight to the point, and this is one topic that I don't think we'll find controversial, so I'm very excited. 

One thing to think about that will make it easier for me to follow:

What is the overall purpose of this document and what do we hope to gain from it

1) This could be a proposal, where we'd like stakeholders including our team and the mobile/releng team to look at, comment, then approve/reject. This would fit best as a google doc of some sort, and an email to the RFC list. Once approved, we commit an ADR and create jira tasks and get going.
2) An informal discussion where we have a general idea, but not yet a concert proposal - so we can discuss and narrow down our thoughts - this could also be a google doc that transforms into the proposal, and maybe we can have an organized meeting specific to this topic, with the concrete goal of coming up with a proposal.
3) A decision record, which could be a better fit for an ADR - where we already accepted/rejected the proposal and now want a long-lasting public record for it so we can always come back

Or is this something separate altogether? Maybe a direction-setting document, where we just want to commit to this direction, then a concrete proposal will follow? think a short sentence of "The purpose of this document is X" will help me contribute better!

lougeniaC64 commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,85 @@
+  * Good, because we don't need to spend time cutting releases.
+  * Good, because uplifts can be done much quicker.
+  * Bad, because we need implement new automation.
+  * Bad, because when breaking changes cause the automatic nightly merge to fail for an application.  This means a developer will get the relatively urgent task of resolving things.
```
Maybe it's better to just back out breaking changes when they happen? The idea that quickly fixing a breaking change (which could be nontrivial) might fall on whoever is around seems undesirable.

bendk replied:
Yeah, that sounds better to me as well.

bendk commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,85 @@
+            * The version string is composed of major, minor, and patch components.  This matches firefox-android, but not moz-central, which skips the minor component.  The reason to have a minor component is to support weekly releases for mobile apps.  They may choose to include new features on those apps, which may require us to uplift new features to our release version and that would be a mi ...
+          * Add a git tag for the commit based on that version number (`v[firefox-version].0.nightly-[YYMMDD]`).
+          * Run the `github-release` CI to publish our code to our consumers (publish maven packages, start the `rust-components-swift` CI, etc).
+        * Automate releases from the `beta`, `release`, and `esr-*` branches whenever a PR is merged
```
- Updated the text to clarify that it's only to the branch.
- I agree this isn't happening that often, so I updated the text to suggest that it's manually handled by relman.

bendk commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,85 @@
+            * Make sure that the github CI can push this commit without needing PR approval.
+            * The version string is composed of major, minor, and patch components.  This matches firefox-android, but not moz-central, which skips the minor component.  The reason to have a minor component is to support weekly releases for mobile apps.  They may choose to include new features on those apps, which may require us to uplift new features to our release version and that would be a mi ...
+          * Add a git tag for the commit based on that version number (`v[firefox-version].0.nightly-[YYMMDD]`).
+          * Run the `github-release` CI to publish our code to our consumers (publish maven packages, start the `rust-components-swift` CI, etc).
```
Yeah, there are lots of details here that this is skipping over to keep the doc simple.  If there's anything that needs discussion, I'm happy to add them back though.

bendk commented:
Thanks for all the great feedback everyone.  I updated the ADR to reflect the comments and also added a section at the bottom that describes the impact to other teams.

This is still a work in progress and nothing is set in stone.  Please interpret the items on the bottom more as questions than anything.  If you don't like how this will impact your team, please add a comment and we can discuss.

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,99 @@
+
+* Status: proposed
+* Deciders: ?
+* Date: ?
```
Should we add today's date as the proposal date?

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,99 @@
+# Rapid releases and moving towards a monorepo
+
+* Status: proposed
+* Deciders: ?
```
What do you think about adding a concrete list of deciders, or at least mention "Relman, Releng, Android, iOS, app services"?

bendk replied:
Sounds great.  I'll start this by putting @rvandermeulen for Relman, and just listing the teams for the others for now.

tarikeshaq commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,99 @@
+
+* **(A) Keep our current release process**
+* **(B) Move to a rapid release process where changes in app-services are frequently merged into consumer applications**
+    * Phase 1: update our branching model and release process
```
Is it feasible to add a timeline to the phases? A rough estimate works too, to give stakeholders an idea of when they should expect the things we list at the bottom

bendk replied:
I put some dates on there based on my guess on how long it would take.  Definitely open to suggestions on this.

tarikeshaq reviewed (commented):
Thank you for adding the section for other teams! Do you think this is ready for an external review? Should we add whoever the deciders are for this as reviewers and publish this to the RFC list?

rvandermeulen commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,106 @@
+* Deciders:
+  * Sync team: bendk, tarikeshaq, lougeniaC64, ddurst
+  * Release management: rvandermeulen
+  * Release Engineering: ?
```
This should be Cyrus Bond

mhammond commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,106 @@
+    * Phase 1: update our branching model and release process (Jan 2023 - Feb 2023)
+        * Update our branches to match how firefox-android works: a `main` branch that is used for the current nightly builds and a `releases_v[N]` for each previous release.
+        * Automate a nightly release of our `main` branch.
+          * Create a commit that updates our version number to something like `[firefox-version].0-nightly-[YYYYMMDD]`, similar to what `prepare-release.py` does now.
```
Do we really need this? I don't think m-c nor nss do this - can we just get away with a date and/or a revision? It's not that unusual for a number of days to pass without a commit.

mhammond commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,106 @@
+          * Create a GitHub release from the branch
+          * Create PRs to merge the changes forward to `main` and more recent release branches.
+        * Continue to maintain our CHANGELOG, but with new versioning.  In main, keep a rolling log of updates with the header `[firefox-version].0`
+        * This replaces our current release process -- no more cutting releases.
```
"no more cutting releases" seems a little misleading here - eg, just below it says "Create a GitHub release..." - maybe just kill the 2nd half of that sentence?

mhammond commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,106 @@
+        * Continue to maintain our CHANGELOG, but with new versioning.  In main, keep a rolling log of updates with the header `[firefox-version].0`
+        * This replaces our current release process -- no more cutting releases.
+        * To push changes to a consumer: create a PR in their repository that bumps the application-services version to an existing release.
+        * On the monthly Firefox merge date:
```
We should automate this as much as possible, even if run manually. So maybe write something like "We will implement a script which will be manually run on the monthly Firefox merge date. This script will:" or similar?

mhammond commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,106 @@
+        * For each consumer application (including all mobile apps and Desktop):
+            * Create a PR to take the new application-services version.
+            * Set up an auto-merge system where if the test suite passes we automatically merge the PR.
+            * Set up an alert that gets sent to `sync-bots` if the test suite fails.  This means we would need to identify why the tests failed and assign a developer to resolve the issue.  The expectation would be that the developer fixes things in time for the next nightly merge.
```
I think there's already a similar process in place for mobile - when we make a new (manual) release, some process kicks in - @rvandermeulen understands this better than I do - maybe link to that as a point of context, just to make it clear there is already something we can beg/borrow/steal for these steps?

isabelrios replied:
For iOS we have a process in place, we check (currently everyday) if there is a new rust-components release. If there is a new one, a [PR](https://github.com/mozilla-mobile/firefox-ios/pull/12649) is created automatically. It pre-builds the app on our CI system (Bitrise) using that new version. We know then if the app will be green or not with that version. Once the job finishes, we send a [slack notification](https://mozilla.slack.com/archives/C4DK7829K/p1670598825686639) the the firefox-ios team

mhammond commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,106 @@
+            * Set up an auto-merge system where if the test suite passes we automatically merge the PR.
+            * Set up an alert that gets sent to `sync-bots` if the test suite fails.  This means we would need to identify why the tests failed and assign a developer to resolve the issue.  The expectation would be that the developer fixes things in time for the next nightly merge.
+    * All application-services changes that are breaking for consumers should have corresponding PRs in the consumer repos that have been reviewed and approved.  If we accidentally merge the change into application-services before the PRs are ready, then we should back the change out.
+
```
Should we explicitly say somewhere that all PRs will automatically attempt to build the mobile browsers? IOW, I'm looking for a process where a PR will not be able to be merged unless we already have a high degree of confidence that it's not breaking, meaning the need to back a change out for this reason should be very rare.

bendk replied:
This is a very good point.  I was thinking that we could merge the PRs one by one, then if there were issues we could make a revert commit and by the next nightly everything would be in sync again.  However, this does seem a bit messy and could cause serious issues if there were database migrations involved.

I updated the previous sections to specify that PRs should be merged "atomically".  No PR should be merged until they all pass CI, then they all get merged more-or-less at once.  For phase 1, they should also all be approved before the merges start.

mhammond commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,106 @@
+
+### Keep our current release process
+
+  * Good, because it doesn't require changes to our workflow
```
We should enumerate what's bad about this - the way this section reads it almost sounds like we should stick with the status quo (status quo is all "good", changed workflow has many "bad" :)

Probably just something like "* Bad, for the reasons enumerated at the top of this document" or similar?

mhammond reviewed (commented):
This looks great to me, thanks! I'll leave formal r+ to the people listed as deciders, but I'm certainly r+ in spirit :)

I think there's still a possibility to remove some of the implementation detail bullet-points - eg, some of my comments below probably wouldn't exist if it was slightly more vague, and I suspect some of the bullet points will not age well as we get more experience. A possibility is to move them to the bottom of the doc into a kind of "non normative discussion about what we think this will look like" - but I'm really not at all bothered by that - do whatever you think is best.

Looking forward to it!

skhamis commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,106 @@
+  * Good, because code will get merged and released more frequently.
+  * Good, because we don't need to spend time cutting releases.
+  * Good, because uplifts can be done much quicker.
+  * Bad, because we need implement new automation.
```
Is this bad? :) I see this as a plus, but I guess more maintenance of those automation processes

skhamis commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,106 @@
+            * Make sure that the github CI can push this commit without needing PR approval.
+            * Note: we will no longer be using semver-style versions and also not following semver semantics.  We will be releasing breaking changes in a nightly cycle without bumping the minor version.
+          * Create a GitHub release from the branch with a tag based on the version number / date (`v[firefox-version].0-nightly-[YYYYMMDD]`). This will cause maven packages to be published, start the `rust-components-swift` CI, etc.
+        * Manually handle releases from the `release_v*` branches whenever new code is committed to those branches
```
I'm a bit confused on what this means -- specifically, "whenever new code is committed". Does this essentially mean we cut when we feel a release is ready? 

Since we will have nighties, feels like we only need these releases to coincide with when the consumers need a stable release (which would probably be at the end of the nightly cycles?).

bendk replied:
I was thinking more about patch releases here.  The "whenever new code is committed" text was trying to say that we should make a patch release right after the merge.  There's no point in having unreleased code sit around in these branches.  Hopefully the next text is more clear about this.

skhamis commented on docs/adr/0005-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,106 @@
+  * Bad, because breaking changes require the corresponding PRs to be approved from all other teams before they can be merged into application-services.
+  * Good, because this is how things will work in a monorepo world.  Adjusting our workflow now prepares us for the future and will give us insights that may help guide the monorepo migration.
+  * Good, because our version numbers will align more closely with the rest of Mozilla.
+  * Bad, because our version numbers will no longer follow semver.
```
Feels like this is not necessarily bad imo -- semver is good for libraries with a variety of consumers that could possibly depend on different versions of a-s (and decide upgrade paths). Where we are essentially a module in various Mozilla products and thus it's better (as you mentioned in the point above) that we're aligning with how we're actually used

bendk replied:
Agreed.  I'll just remove this item, I think the note that we're breaking with semver is all we need.

bendk commented:
This feedback is great, thanks all.  I pushed a change that:

* Adds a deadline for feedback
* Moves the details about the nightly version to a section at the bottom
* Specifies that all PRs should be merged together
* Addresses the various other feedback.  I marked a bunch of conversations resolved, since hopefully I clarified all the points.  Feel free to re-open any discussions though.

RE nightly versions: The reason I suggested making releases for these, with new version numbers, etc. was that we would get some automation for free, since it would work with the existing automation we have with iOS and Android.  But I don't have strong feelings and I agree that this is more of an implementation detail than anything.

csadilek commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,122 @@
+
+Mozilla generally favors a rapid release model to long cycle times.  Firefox desktop adopted rapid releases over a decade ago and there is current work to move both the mobile applications and our services towards rapid releases.  Although there are some costs to rapid releases, as a company we've decided that we're willing to accept them.
+
+Application services should consider moving towards a rapid release model.  The same reasoning that led other teams to adopt rapid releases also applies to us.  Furthermore, the mobile teams are moving towards rapid releases which may depend on us also adopting them.  Finally, we are in the process of moving towards a single monorepo for android code, with potential to add iOS code and/or move in ...
```
One bit that confused me here. You write that Firefox Desktop adopted rapid releases over a decade ago, and mobile is moving toward it. Mobile is already on pretty much the same release cadence as Desktop, or did this specifically refer to iOS, where the cadence is the same, but Nightly|Beta is TestFlight only?

csadilek replied:
(The ongoing rapid release project aims to increase cadence significantly relative to that...on mobile).

bendk replied:
Good point, I think I got confused by discussions of the "new rapid release model".  I'm going to rewrite this to say:
  - Mobile has more-or-less the same release cadence as desktop
  - There's work going on to speed it up even more.

csadilek commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,122 @@
+        * Update our branches to match how firefox-android works: a `main` branch that is used for the current nightly builds and a `releases_v[N]` for each previous release.
+        * Automate a nightly release of our `main` branch
+          * See the discussion at the bottom of this document for how this could work
+          * Note: regardless of which option we choose, we will no longer be using semver-style versions and also not following semver semantics.  We will be releasing breaking changes in a nightly cycle without bumping the major version.
```
It would be great to call out that A-S would then follow the same versioning as all of M-C and mobile, which is a big win, in my opinion! We will no longer need to ask "but, which version of A-S is Fenix 110 using?"

bendk replied:
Yeah it'll be nicer for sure.  I added a pro in the list for this.

Note: I don't think we'll be able to keep the minor versions in sync between a-s and Fenix, since Fenix may want to release a minor version without any app-services changes.  Also, I think Fenix, iOS, and Desktop don't keep their minor versions in sync, which also would prevent this.  So I guess there will still be some digging needed for previous minor version releases, but it should be much simpler than the current situation.

rvandermeulen replied:
Thinking about this more (especially with a longer-term view of A-S potentially living in m-c some day with a separate-but-related version to the regular version.txt), we should probably have a test somewhere which enforces that the A-S major version matches the Gecko one.

nbhasin2 replied:
Hi, what would be the distinguishing factor for a release made for Firefox Android vs iOS when we generate dot releases?

Is the idea here that we put out a dot release that will basically work for all platforms but might have improvements/bugfixes for the specific platform it is intended for?

bendk replied:
Yes the idea is that when we did a dot release it would be for all platforms at once (Desktop too).

There's probably some scenarios when a particular app wouldn't take the new dot release.  For example, it was for a component that's not used.  But I think this should be pretty rare.  Even in the case where it's not going to affect the app, if the app was going to make a dot release anyways then I think they might as well take the latest app-services release.

rvandermeulen replied:
I think there's a distinction to be made between an app "taking" a new dot release as in "the PR was made, auto-merged, is effectively a no-op and will ship whenever it happens to" and "the PR was merged and we shipped a new release off of it".

csadilek reviewed (commented):
Thanks for the ping and working through this! Just want to voice my support for this change. I think aligning our release cadences will make our product more stable and predictable in the long run. It will also simplify release processes, as well as development.

Detecting breaking changes early, and no longer having to ask which version of A-S is being used in given Firefox release/branch alone will be a big win! So, +1 from me :).

bendk commented:
> Thanks for the ping and working through this! Just want to voice my support for this change. I think aligning our release cadences will make our products more stable and predictable in the long run. It will also simplify release processes, as well as development.
> 
> Detecting breaking changes early, and no longer having to ask which version of A-S is being used in given Firefox release/branch alone will be a big win! So, +1 from me :).

Thanks for taking a look at this!  I added you as the decider for android and put a check mark next to your name.  Hopefully I'm not jumping the gun here, feel free to tell me to remove it if you have any concerns.

tarikeshaq reviewed (approved):
Official +1 for me!

ddurst commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,123 @@
+  * Release Engineering: cyrus-bond
+  * Firefox Android:  ✅ csadilek
+  * Firefox iOS: ?
+  * Firefox Desktop: ?
```
Cyrus Bond and Sylvestre can be added for desktop. I'll email them to make them aware -- they may want to add others.

rvandermeulen commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,123 @@
+
+* Phase 1
+  * The application-services version number style will change
+  * There may be a increase in the amount of application-service version upgrade PRs to approve.  We'll want to make sure that each application is on an application-services release that matches their release (i.e. avoiding having Fenix 110.x on application-services 109.x).  For the Android/iOS teams this will only be a small increase, but for Desktop it will be larger.
```
Can we leverage updatebot for this on the Desktop side to automate most of the process? So eventually someone on the A-S team just needs to r+ a Phabricator rev and queue it for Lando?

bendk replied:
That sounds pretty great to me.  I'm hoping we can leverage the existing automation as much as possible.

rvandermeulen replied:
Feel free to resolve this again, but I also just wanted to draw your attention to the NSS process so you can see another option. They utilize a single tracking bug for every cycle (for example: https://bugzilla.mozilla.org/show_bug.cgi?id=1805486) with periodic pushes during the Nightly cycle until the final release is tagged at the end of it. For A-S, I think I'd lean towards preferring having a new bug for each upstream bump (bugs and automation are cheap after all, and it makes for better bisection down the road), but that's another option anyway.

rvandermeulen commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,123 @@
+  * Will now handle tagging, releasing, and version bumping application-services at the end of the nightly cycle.
+  * Will now handle patch releases from the release / nightly branches
+* Phase 2
+  * Maybe handle backing out changes when the auto-merge fails?  I'm not sure who should own this.
```
I'm not clear what this is saying, but in general I'd expect developers to be responsible for backing out changes which cause problems. As an organization, we've been pretty clear that only our Gecko mercurial repos get proactive sheriffing coverage.

bendk replied:
Seems fair to me, I'll remove that.

rvandermeulen commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,123 @@
+* Phase 2
+  * More application-services PRs to approve, especially at the start.  Eventually, we should be able to arrange it so most PRs get auto-merged.
+
+### Release Management
```
I've very hesitant to sign off on our team taking this on without having a better understanding of what the workflow would be here. It sounds like these steps are intended to be mostly automated per the plan above? Is this just going to be a matter of triggering a hook somewhere, starting a build in ship-it, or something else?

rvandermeulen replied:
And from an expectations standpoint, for releases, is the expectation that a new point release will be created for every uplift to a release branch or done on some schedule? Or when requested by the team?

bendk replied:
> I've very hesitant to sign off on our team taking this on without having a better understanding of what the workflow would be here. It sounds like these steps are intended to be mostly automated per the plan above? Is this just going to be a matter of triggering a hook somewhere, starting a build in ship-it, or something else?

I don't think we know exactly how this is going to work, and I wanted to leave those details out of the ADR, but  it should certainly be automated.  The work should be something like triggering a hook and monitoring that it succeeds.  I wasn't sure if that should be our team's responsibility or relman, I'm happy to follow your recommendation here.

> And from an expectations standpoint, for releases, is the expectation that a new point release will be created for every uplift to a release branch or done on some schedule? Or when requested by the team?

With our current workflow, we backport fixes about once a month or less.  Based on that, I think we should make a new point release as soon as the uplift is merged since it's unlikely that any more uplifts are coming soon.

rvandermeulen replied:
We can take this discussion outside the ADR if you want. But my high-level recommendation would be to utilize Ship-It for creating releases and have those be self-serviced by the A-S team. RelMan would then be involved as usual with approving uplifts to the appropriate release branches after the new release is ready. We can talk through the details in greater depth later, but this has some strong advantages over manually tagging releases in GitHub or triggering hooks in TC to kick off a new release and aligns well with the overall direction we're trying to move in around consolidating releases through there.

rvandermeulen commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,123 @@
+The ADR mentions making nightly releases from the main branch.  There are two basic approaches for this:
+
+  * **Create new releases**: Create a new version number, GitHub tag, GitHub release, Kotlin/swift packages, etc.
+  * **Update a single release**.  Use a name like `[firefox-version].0a1` for the nightlies and don't change it.  Don't create GitHub tags or releases.  Re-publish the existing Kotlin/Swift packages rather than create new ones.
```
IMO, Nightlies shouldn't be getting tagged as releases. I would think we'd be doing something akin to how we ship new GeckoView nightlies every day where they get uploaded to the Nightly Maven repo with a [version].[date] numbering scheme, i.e. 110.0.20230105143224 as shown in https://nightly.maven.mozilla.org/?prefix=maven2/org/mozilla/components/browser-engine-gecko/. Then the first tagged release for a given train would be after the .0 release is branched and shipped.

bendk replied:
This makes a lot of sense to me as well.  The only reason for tagging them is that the existing automation creates the PRs when there's a new release, so maybe it simplifies the implementation.  Hopefully we can modify the automation to do what you suggest though.

bhearsum replied:
I'll add that no tags or releases for nightlies is consistent with what we do elsewhere, too. (Eg: Fenix uses [Taskcluster's cron scheduling](https://github.com/mozilla-mobile/fenix/blob/main/.cron.yml) for nightlies.)

bendk commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,123 @@
+        * This replaces our current release process.
+        * To push changes to consumers:
+           * Create a PR in their repository that bumps the application-services version to one of these releases.
+           * Don't merge any of these PRs unless all pass CI and are approved.
```
I think that's one of the big questions here.  In general, I think yes it should.  Otherwise we run into the situation where iOS is behind Android and therefore can't take any new code at all.  If there's a second breaking change that comes, it's very hard to resolve it when the first breaking change is still unresolved.  If there's an important bug that needs to be addressed then we're in a very bad situation -- even worse than the current situation, since we can't backport the fix.

Maybe we could allow some exceptions, but I think they would need to be really strongly motivated.

What do you think?

rvandermeulen replied:
To me, that concern can be mostly solved by having a strong "no bustage" policy where broken commits are aggressively backed out in the event that a breaking change lands. This is effectively what we already enforce, for example, with Gecko changes that break Fenix. My belief is that in a world where there's ongoing Nightly builds landing on a regular cadence, this becomes a more workable policy as the feedback loop becomes a lot tighter and more manageable.

JohanLorenzo replied:
I agree with @rvandermeulen's proposal: no bustage policy. It'll be easier to automate PRs being merged if one PR doesn't have to wait on another one on a different repo.

bendk commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,123 @@
+          * Bump the major version number in main and start a new changelog header.
+    * Phase 2: Automate merges of `application-services` into consumer applications (March 2023 - May 2023)
+      * Every night, for each consumer application (including all mobile apps and Desktop):
+          * Create a PR to take the current application-services code in `main`.
```
Yeah and iOS has something like this as well.  I'm hoping that we can leverage those as much as possible.

bendk commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,123 @@
+    * Phase 2: Automate merges of `application-services` into consumer applications (March 2023 - May 2023)
+      * Every night, for each consumer application (including all mobile apps and Desktop):
+          * Create a PR to take the current application-services code in `main`.
+          * Set up an auto-merge system to automatically merge the PRs.  As with phase 1, no PR should be merged until all PRs pass CI and then all PRs should be merged at once.
```
> Who's going to be in charge of watching these PRs across every repo and manually merging them all at the same time? 

I was envisioning some automation that did that, but that might be magical thinking since I don't know how it would be implemented.

The alternative would be to auto-merge the PR for each repo if it passes the CI for that repo, without paying attention to the other CI.  That has a very clear implementation path, just extend the existing integrations to auto-merge.

The issue is that maybe the CI passes for one repo, but not another, and they get out of sync.  In general, that's not so bad and usually the worst case scenario is that you just have to revert the change and everything is fixed in the next nightly.  However, sometimes that won't be possible, for example if a change contains a DB migration.

I really like your suggestion of auto-merging without checking the other repos, but having a PR flag to prevent that behavior and require manual merging for the special cases where we can't simply revert the code.  What do others think about this?

rvandermeulen replied:
As mentioned above, we can couple auto-merges with a strict backout policy for breaking changes to ensure that things don't remain out of sync for long when they do break.

rvandermeulen commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,121 @@
+           * Don't merge any of these PRs unless all pass CI and are approved.
+           * Once the PRs are approved, then merge them all at once.
+        * We will implement a script which will be manually run on the monthly Firefox merge date. This script will:
+          * Update the version number to be the final release version, e.g. change `v[N].0a1` to  `v[N].0`
```
Minor nit - this step should come after the branch is created

bendk commented:
Just updated the deadline from yesterday to next Monday, the 16th.  The initial deadline was more of a guess than anything, but I'm hoping we can stick to the the new deadline and wrap this up by the 16th.

bhearsum commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,121 @@
+          * Create a GitHub release for that branch
+          * Bump the major version number in main and start a new changelog header.
+    * Phase 2: Automate merges of `application-services` into consumer applications (March 2023 - May 2023)
+      * Every night, for each consumer application (including all mobile apps and Desktop):
```
Is the intention that these would merged every day, or do you want daily PRs more to make sure you don't break anything in the applications?

bhearsum commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,121 @@
+        * This replaces our current release process.
+        * To push changes to consumers:
+           * Create a PR in their repository that bumps the application-services version to one of these releases.
+           * Don't merge any of these PRs unless all pass CI and are approved.
```
Slightly tangential to Ryan's point -- what is the benefit of waiting for, eg: the PR in firefox-ios to be approved before merging an update for fenix? I can certainly see the benefit of waiting for CI to pass in all repos, but I'm having trouble understanding why we'd want to wait on review everywhere.

bendk replied:
What happens now is that we get one PR approved and merged, say on Android, but then the iOS PR can sit around for a long period of time before being approved.  The reasons vary: maybe it's a large change on the iOS side, maybe they want to think through changes that affect the UX, maybe they just don't have the bandwith to review something.

In any case, what ends up happening that Android and iOS have very different versions of app-services, which can get confusing.  Also one of them is blocked from taking any new code until the PR is reviewed.

bhearsum commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,121 @@
+  * Good, because code will get merged and released more frequently.
+  * Good, because we don't need to spend time cutting releases.
+  * Good, because uplifts can be done much quicker.
+  * Good, because more of our release process will be automated.
```
This is a bit theoretical, perhaps, but another benefit of aligning with the same process that Firefox, et. all use is that you may benefit later from tools we build for other things. For example, you may want to integrate with [Buildhub](https://buildhub.moz.tools/) or [Product Details](https://product-details.mozilla.org/1.0/) someday - and that's easier or maybe even free when your process is the same.

bendk replied:
Good point, I'll add this one.

bhearsum commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,121 @@
+  * Good, because uplifts can be done much quicker.
+  * Good, because more of our release process will be automated.
+  * Bad, because we need to implement new automation.
+  * Bad, because when breaking changes cause the automatic nightly merge to fail for an application.  This means a developer will get the relatively urgent task of resolving things.
```
Is this negated by Ryan's encouragement to have a strict no-bustage/backout policy?

bendk replied:
I'm assuming that we'll mess that up at some point and need to fix things.  But I think it makes sense to add a sentence mentioning that this will only happen if we don't follow the policy properly.

bhearsum commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,121 @@
+  * Bad, because we need to implement new automation.
+  * Bad, because when breaking changes cause the automatic nightly merge to fail for an application.  This means a developer will get the relatively urgent task of resolving things.
+  * Good, because this avoids the current state of affairs where our `main` branch is not ready to be taken by the applications.  This means that a) our current code is not getting used and tested and b) it's very complicated to get other changes to that application, including bugfixes.
+  * Bad, because breaking changes require the corresponding PRs to be approved from all other teams before they can be merged into application-services.
```
I know I'm missing something about the process here...but how does this work? Doesn't the breaking change need to be landed and released by application-services before consuming applications can pick up and make their changes?

bendk replied:
Yeah, I think we'll need to create a way to run the iOS/Android/Desktop CI using an in-progress app-services branch.

For Desktop, I think it would be relatively simple to automate vendoring in our code then running a try build.

For Android/iOS, there's currently code to build/test against an app-services branch, but it needs to be run manually.  the task would be to get it to run in CI for a PR.

I'll add an item in the plan for this.

ahal commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,121 @@
+
+## Context and Problem Statement
+
+Application services currently has a long cycle time for code changes.  We make a release from main on an ad-hoc schedule, averaging about 3-4 releases a month.  Our consumer applications take our releases also on an ad-hoc basis, `firefox-android` takes our releases fairly frequently, `firefox-ios` less frequently, and desktop very infrequently.  This means that code changes take a relatively lo ...
```
A bit of a noob question, but I have no idea how app-services works :)..

This paragraph talks about different products taking app-services releases at different frequencies. However most of the rest of the doc seems to be centered around the problem of needing to co-ordinate PRs to all land at the same time (presumably so they're all using the same version?). Those seem contradictory to me, do all products need to be using the same version at once?

bendk commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,121 @@
+          * Set up an alert that gets sent to `sync-bots` if the test suite fails.  This means we would need to identify why the tests failed and assign a developer to resolve the issue.  The expectation would be that the developer fixes things in time for the next nightly merge.
+          * Note: we currently have automation to do some of this for both iOS and Android.  We should leverage the existing code as much as possible.
+      * Perform the same process whenever a new application-services version is published in a `release_v[N]` branch.
+    * All application-services changes that are breaking for consumers should have corresponding PRs in the consumer repos that have been reviewed and approved.  If we accidentally merge the change into application-services before the PRs are ready, then we should back the change out.
```
This is referring to scenarios like this:
   - We create an app-services PR with breaking changes
   - The PR to resolve those changes are approved on Android, but not approved for on iOS for some reason
   - We merge the app-services PR and the android PR.
   - We figure out that the iOS PR is not ready and the iOS devs want to take some more time to review it before approving.
   - At this point, we should make commits to revert the app-services changes (as well as the corresponding Android changes as well).  This is in contrast to our current practice which would be to just let the Android a-s version get out of sync from the iOS a-s version.

JohanLorenzo commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,123 @@
+
+### Release Engineering
+
+* Support the application-services team in developing the automation.  I believe most of the proposed changes are simple enough that application-services team can handle it ourselves.
```
Just for the record, I know for sure Releng has this knowledge on Android. If I'm not mistaken another team dealt with this type of automation on Desktop (maybe the folks handling Lando?). I don't know who would be the right people to help with Firefox iOS. Because of the Apple ecosystem is mainly closed, the CI/CD is not centered on Taskcluster there.

JohanLorenzo reviewed (approved):
Thank you for writing this record! I'm pretty excited about the result! 

I agree with the "no bustage" policy @rvandermeulen suggested. I think it's going to be easier for Mozillians of all teams to remember "Platform X is temporarily out of date because of Y", rather than "Lib X is out of date everywhere because of Y on another platform".
To me, the main reason is: that's what most Mozillians are used to. For instance, we keep providing Windows Firefox nightlies although Linux ones are busted, or vice-versa. The same rationale occurs between Desktop and Android.

bendk commented:
Thanks for the review.  I think I'm understanding the "no bustage" policy now and pushed some changes based on that.  It seems to me that we have basic agreement here and it's time to approve this and start working on it.  Are there any concerns that still need to be addressed?

sylvestre commented on docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md:
```diff
@@ -0,0 +1,126 @@
+  * Release Engineering: jlorenzo
+  * Firefox Android:  csadilek
+  * Firefox iOS: nish
+  * Firefox Desktop: cyrus-bond, sylvestre
```
```suggestion
  * Firefox Desktop: sylvestre
```

bendk commented:
Thanks for all the help with this one.

It looks like there's broad agreement on this, so I'm going to merge it and start figuring out how to implement this.
